vault/init.go

```diff
@@ -270,54 +271,63 @@ func (c *Core) Initialize(ctx context.Context, initParams *InitParams) (*InitRes
 	return results, nil
 }
 
-// UnsealWithStoredKeys performs auto-unseal using stored keys.
+// UnsealWithStoredKeys performs auto-unseal using stored keys. An error
+// return value of "nil" implies the Vault instance is unsealed.
+//
+// Callers should attempt to retry any NOnFatalErrors. Callers should
+// not re-attempt fatal errors.
 func (c *Core) UnsealWithStoredKeys(ctx context.Context) error {
+	c.unsealWithStoredKeysLock.Lock()
+	defer c.unsealWithStoredKeysLock.Unlock()
+
 	if !c.seal.StoredKeysSupported() {
 		return nil
 	}
 
 	// Disallow auto-unsealing when migrating
 	if c.IsInSealMigration() {
-		return nil
+		return NewNonFatalError(errors.New("cannot auto-unseal during seal migration"))
 	}
 
 	sealed := c.Sealed()
 	if !sealed {
+		c.Logger().Warn("attempted unseal with stored keys, but vault is already unsealed")
```

Comment:
Given that the goroutine runs until shutdown occurrs, won't this continually spam logs?

Reply:
The goroutine halts the first time `UnsealWithStoredKeys` returns nil (or when a fatal error occurs), so this log message would only print once if the Vault was already unsealed.

So it'd be:

1. go func
1. UnsealWithStoredKeys
1. Log message
1. go func terminate

There _will_ be log spam when a Vault server is brought online for the first time and the storage backend isn't yet initialized, but that's already the case.

Reply:
Oop yes you're right. I had thought you intended for it to run always in case it seals (which I wasn't sure was a good thing) so figured that'd get hit over and over. Sounds good!